Iterate on launch stats, to include tracking for all and first-time launches

SiobhyB opened the pull request:
Related to 7578-gh-Automattic/dotcom-forge

## Proposed Changes

- This PR includes the following proposals for the tracking we have around app launches:
   - Add code comments around the current stat, `local-environment-launch-uniques`, to convey its purpose. A separate issue will be created to potentially rename the stat in the future.
   - Introduce a new stat, `studio-app-launch-total`, that fires every single time the app is launched.
   - Introduce `local-environment-launch-uniques-first`, which will fire first time a unique user first launches the app.

>  [!NOTE]
> I'm keen to hear thoughts around naming. I've spent a bit of time looking closely at this problem, so newer eyes will likely better spot any names that could be made clearer.

## Testing Instructions

- Apply the following patch to ensure the stats are bumped in your local development environment, and that they're logged to the console when they fire:

```patch
diff --git a/src/lib/bump-stats.ts b/src/lib/bump-stats.ts
index 8154cec..05bd77c 100644
--- a/src/lib/bump-stats.ts
+++ b/src/lib/bump-stats.ts
@@ -49,12 +49,15 @@ export function bumpAggregatedUniqueStat(
 }
 
 // Returns true if we attempted to bump the stat
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
 export function bumpStat( group: string, stat: string, bumpInDev = false ) {
-	if ( process.env.E2E || ( process.env.NODE_ENV === 'development' && ! bumpInDev ) ) {
+	if ( process.env.E2E ) {
 		console.info( `Would have bumped stat: ${ group }=${ stat }` );
 		return false;
 	}
 
+	console.log( `Bumped stat: ${ group }=${ stat }` );
+
 	const url = new URL( 'https://pixel.wp.com/b.gif?v=wpcom-no-pv' );
 	url.searchParams.append( `x_${ group }`, stat );
 

```
- In order to mimic a first launch, remove the `Studio` folder found under `Library` → `Application Support`.
- Follow the steps to launch the Studio app. 
- Ensure the following are logged to your console when first launched:
  - `Bumped stat: studio-app-launch-total=darwin`
  -  `Bumped stat: local-environment-launch-uniques=darwin`
  -  `Bumped stat: local-environment-launch-uniques-first=darwin`
- Exit the app and launch it a second time.
- Verify that _only_ the following stat is bumped on the second launch:
  - `Bumped stat: studio-app-launch-total=darwin`

>  [!NOTE]
> An MC page for each stat will be automatically created, but you may not see your stat reflected immediately.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Have you checked for TypeScript, React or other console errors?

SiobhyB committed:
feat: Track the first occurrence of a launch

SiobhyB committed:
feat: Track every single app launch

SiobhyB committed:
refactor: Rename weekly stat for clarity

`local-environment-launch-uniques` has been updated to `studio-app-launch-weekly-uniques` in a bid to clarify its purpose, which felt especially needed with the recent introduction of `studio-app-launch`.

SiobhyB committed:
docs: Add code comments to clarify purpose of each new stat

SiobhyB commented on src/index.ts:
```diff
@@ -228,7 +228,10 @@ async function appBoot() {
+		// Bump a stat on each app launch
+		bumpStat( 'studio-app-launch', process.platform );
+		// Bump stat for unique weekly app launch, approximates weekly active users
+		bumpAggregatedUniqueStat( 'studio-app-launch-weekly-uniques', process.platform, 'weekly' );
```
I'm very open to naming suggestions here. I went with a rename as it felt like the old name would cause confusion with the fact we're now going to track every launch separately.

Note, as a follow up to this PR, we'll need to merge the `local-environment-launch-uniques` stats into the new name. There is a dedicated page with guidance for this in the FG.

SiobhyB commented on src/lib/bump-stats.ts:
```diff
@@ -20,7 +20,10 @@ export function bumpAggregatedUniqueStat(
-				return bumpStat( group, stat, bumpInDev );
+				bumpStat( group, stat, bumpInDev );
+				// Track the first occurence of this stat separately
+				bumpStat( `${ group }-first`, stat, bumpInDev );
```
This felt like the simplest approach to tracking first-time launches. 

I also experimented with a separate `bumpFirstUniqueStat` function that worked very similarly to `bumpAggregatedUniqueStat`. However, `bumpFirstUniqueStat`  and `bumpAggregatedUniqueStat` did not play nicely together due to the async nature of [`updateLastBump`](https://github.com/Automattic/studio/blob/fed78eb88bb3b91ae37255c94d1a673d681a5e2c/src/lib/bump-stats.ts#L73-L79). Only one would fully execute on the first launch, while the second would be fire a second time when the app was launched again. If we wanted to go with a second `bumpFirstUniqueStat` function, I believe we'd need to refactor the existing functions to use `async/await`, which this comment led me away from doing:

https://github.com/Automattic/studio/blob/fed78eb88bb3b91ae37255c94d1a673d681a5e2c/src/lib/bump-stats.ts#L12-L13

I'd definitely be interested to hear any concerns on the approach I opted for, or other suggestions!

wojtekn replied:
Would it make sense to add a comment above each of the `bumpStat( group, stat, bumpInDev );` calls to clearly mark that one is for the first launch, and the second one for subsequent launches in the next interval?

SiobhyB replied:
Good idea, I went ahead to add comments in https://github.com/Automattic/studio/pull/221/commits/48041a41d442692d4705ead3fd419952b4285d79. There are also comments in the `index.ts` file:

https://github.com/Automattic/studio/blob/48041a41d442692d4705ead3fd419952b4285d79/src/index.ts#L231-L234

Let me know if you think the wording should be tweaked or made clearer in any way!

wojtekn replied:
I think it's clear now, thanks for adding those comment.

SiobhyB committed:
test: Add mocking for stat with `-first` suffix

danielbachhuber commented:
Will this break our existing stat? If so, I'd prefer to keep the existing naming so we have the continuity of data in place.

SiobhyB commented:
> Will this break our existing stat? If so, I'd prefer to keep the existing naming so we have the continuity of data in place.

@danielbachhuber, it wouldn't break the existing stat, no. There's a process for renaming MC stats outlined in the FG, though we'd just need to be careful about timing if we go ahead with that. I've added a `[Status] Not Ready for Merge` label so folks know not to merge this PR before we're ready.

I'm also keen to hear thoughts or other suggestions on the naming of the stats. I obviously don't want to get in the way if the current weekly stat name works, but now feels like a good opportunity for reviewing and making sure it's as clear as possible.

danielbachhuber commented:
@SiobhyB Sounds good, thanks for clarifying

wojtekn commented:
> I've added a [Status] Not Ready for Merge label so folks know not to merge this PR before we're ready.

If we have a process for renaming, it would be good to do so, as it would keep those names consistent and avoid confusion. However, as Studio is an app with a release schedule, and there is a possibility that users won't upgrade it, we may have both old and new stats being logged in parallel. Will the rename still work in our case?

SiobhyB commented:
> However, as Studio is an app with a release schedule, and there is a possibility that users won't upgrade it, we may have both old and new stats being logged in parallel. Will the rename still work in our case?

@wojtekn, that's a good point, my assumption was that we'd be able to set up a redirect for the old `http://pixel.wp.com/b.gif...` URL, but I'll ask about to confirm whether this is possible and follow up.

SiobhyB committed:
docs: Add comments to communicate reason for bumps

SiobhyB committed:
refactor: Rename stats to fit tech limits

As there isn't a straightforward way to rename `local-environment-launch-uniques` without retaining previous data, this commit reverts the previous renaming. `studio-app-launch` has also been tweaked to `studio-app-launch-total`, to further distinguish it and hopefully improve clarity.

SiobhyB commented:
@wojtekn, it seems that the most straightforward approach to renaming would be to keep going back to merge the old stats. As I can't commit to this at the moment, due to my upcoming sabbatical, I've gone ahead to revert back to the old name in https://github.com/Automattic/studio/pull/221/commits/5e9a33c8fef39dca7c7430af49839cf610b39bd0. So, now this PR proposes the following:

* `local-environment-launch-uniques`: Tracks unique weekly app launches, approximates weekly active users.
* `local-environment-launch-uniques-first`: Tracks the first time the app is launched by a user.
* `studio-app-launch-total`: Tracks every single time the app is launched (I further tweaked the name in the last commit to further differentiate from the existing stats).

Looking forward to hearing throughts!

I'll also create a separate GitHub issue to detail what needs to be done to rename `local-environment-launch-uniques`, in case someone else would want to pick that up. I see we can add descriptions at the top of the MC stat pages, so will do that too.

Slack discussion related to renaming for context: p1718037250491319-slack-C029GN3KD

danielbachhuber commented:
I'm personally fine with the separate naming. I don't think it's worth the engineering effort to try to standardize.

wojtekn commented:
Thanks, @SiobhyB, for exploring the possibilities. I agree with @danielbachhuber's assessment that it makes more sense to keep that simple with inconsistent naming when compared to maintaining the name change over time.
